_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+### Testing with Metasploit Module Loader
+
+The build process was complicated at the time,
+
+Build metasploit_data_models 0.0.2.43DEV for pro:
+
+{% highlight sh %}
+cd ~/git/rapid7
+git clone git@rapid7.github.com:rapid7/metasploit_data_models.git
+cd metasploit_data_models
+git checkout dd6c3a3
+gem build metasploit_data_models_live.gemspec
+{% endhighlight %}
+
+Build prototype_legacy_helper for pro:
+
+{% highlight sh %}
+cd ~/git
+mkdir jvennix-r7
+cd jvennix-r7
+git clone git@rapid7.github.com:jvennix-r7/prototype_legacy_helper.git
+cd prototype_legacy_helper
+vim prototype_legacy_helper.gemspec
+{% endhighlight %}
+
+I don't remember how prototype_legacy_helper was actually built and installed for production, so let's just use a fake gemspec:
+
+{% highlight ruby %}
+# coding: utf-8
+lib = File.expand_path('../lib', __FILE__)
+
+Gem::Specification.new do |spec|
+  spec.name          = "prototype_legacy_helper"
+  spec.version       = "0.0.0"
+  spec.authors       = [""]
+  spec.email         = [""]
+  spec.summary       = ""
+  spec.description   = ""
+  spec.homepage      = ""
+  spec.license       = "MIT"
+
+  spec.files         = `git ls-files -z`.split("\x0")
+  spec.executables   = spec.files.grep(%r{^bin/}) { |f| File.basename(f) }
+  spec.test_files    = spec.files.grep(%r{^(test|spec|features)/})
+  spec.require_paths = ["lib"]
+
+  spec.add_development_dependency "bundler", "~> 1.7"
+  spec.add_development_dependency "rake", "~> 10.0"
+end
+{% endhighlight %}
+
+Build the gem
+
+{% highlight sh %}
+gem build prototype_legacy_helper.gemspec
+{% endhighlight %}
+
+Rewind pro and metasploit-framework to commits where pro had ActiveSupport::Dependencies, but framework was still incompatible.  (**NOTE: this state is simulated.  When it happened, framework was updated first before pro was committed, but this state did occur on my dev machine while working on this change.**)
+
+{% highlight sh %}
+cd ~/git/rapid7
+git clone git@rapid7.github.com:rapid7/pro.git
+cd pro
+git checkout  3370204
+rm msf3
+ln -s ~/git/rapid7/metasploit-framework ms3
+cd msf3
+git checkout 8b8da0b
+cd ..
+cd ui
+rvm use ruby-1.9.3@pro
+edit Gemfile
+# Comment out therubyracer.  The locked version can't be built against modern Yosemite
+RAILS_ENV=development bundle install
+{% endhighlight %}
+
+So, everything loads using autoload paths and Metasploit Pro developers no longer have to worry about forgetting to adding requires to prosvc to shared code between `ui` and `engine`, right? Trigger the need to autoload `SocialEngineering` using autoload_paths by removing the constant and then trying to use web_phish module that references `SocialEngineering::Campaign`.  Nope, I get "wrong constant name" when trying to load certain modules for the searching:
```

Comment:
So, everything loads using autoload paths and Metasploit Pro developers no longer have to worry about forgetting to **~~adding~~** **add** requires to prosvc to shared code between `ui` and `engine`, right? Trigger the need to autoload `SocialEngineering` using autoload_paths by removing the constant and then trying to use web_phish module that references `SocialEngineering::Campaign`.  Nope, I get "wrong constant name" when trying to load certain modules for the searching: